csrc/pos_encoding_kernels.cu

```diff
@@ -77,6 +77,48 @@ __global__ void rotary_embedding_kernel(
   }
 }
 
+template<typename scalar_t, bool IS_NEOX>
+__global__ void batched_rotary_embedding_kernel(
```

Comment:
This kernel is almost exactly the same as `rotary_embedding_kernel` and you can make them the same by adding the `const int64_t* __restrict__ cos_sin_cache_offsets` argument there and then down below, doing
```
int64_t cos_sin_cache_offset = cos_sin_cache_offsets ? cos_sin_cache_offsets[token_idx] : 0;
```

Reply:
cos_sin_cache_offset is passed as a pointer, we don't have a good way to determine if it's empty without auxiliary flag, also we try to avoid runtime branching in kernel code for performance. agreed that these two kernels are pretty much the same to I refactored it to avoid too much code duplication.